### PR TITLE
Extended Can

### PR DESCRIPTION
**Pull request by save-buffer**
Cai needs to change CAN initialization to have an option to make Extended CAN optional. Currently the call to CANBBB.Initialize() in BBBPinManager.cs gives can0 true (Extended CAN Enabled) and can1 false (Extended CAN disabled).

### CODE REVIEW

File: IO/BeagleBone/BBBPinManager.cs [hunk 1]
```diff
@@ -9,1355 +9,1367 @@
...
-                    Output.Add("    ");
-
-                    if (I2CDev1.Count > 0)
-                    {
-                        Output.Add("    fragment@20 {");
-                        Output.Add("        target = <&i2c1>;");
-                        Output.Add("        __overlay__ {");
-                        Output.Add("            status = \"okay\";");
-                        Output.Add("            pinctrl-names = \"default\";");
-
```
Comment (Baldstrom) [inline IO/BeagleBone/BBBPinManager.cs]: Do we intend to enable `bus0` by default?

Comment (Baldstrom) [inline IO/BeagleBone/BBBPinManager.cs]: Is this _and_ line 565 necessary?

File: IO/BeagleBone/CANBusBBB.cs [hunk 2]
```diff
@@ -152,15 +236,29 @@ internal CANBusBBB(string CanName)
 		/// <param name="Data">Payload of CAN Frame. Must be at most 8 bytes.</param>
```
Comment (Baldstrom) [inline IO/BeagleBone/CANBusBBB.cs]: Typically in our documentation we have `<tag> Description </>` (note the spaces). Also, if we can reasonably make the `<summary>` tag a one-liner, we try to do so. This happens a couple times in this PR. If you could fix this in all the relevant files, that would be great :)

File: IO/BeagleBone/CANBusBBB.cs [hunk 3]
```diff
@@ -86,6 +86,30 @@ private unsafe struct CANFrame
...
+			[FieldOffset(7), MarshalAs(UnmanagedType.U1)]
+			public byte Reserved1;
+
+			//The original source forced this to 8-byte alignment with __attribute__((aligned(8)))
+			//If something breaks, maybe fix it?
+			[FieldOffset(8), MarshalAs(UnmanagedType.ByValArray, SizeConst = 64)]
+			public fixed byte Data[64];
+		}
+
 		[DllImport("libc", SetLastError = true)]
```
Comment (Baldstrom) [inline IO/BeagleBone/CANBusBBB.cs]: In the past, we have separated our `DllImport`s into different files (see our RPi I/O devices). We could consider doing the same here.

File: IO/BeagleBone/BBBPinManager.cs (hunk 1 above)
Comment (CaiB) [inline IO/BeagleBone/BBBPinManager.cs]: Without knowing what was changed, we can't merge this.
You'll need to fix whatever is causing this entire file to be one changeset in order for me to be able to review.

File: Components/Sensors/VH400.cs [hunk 4]
```diff
@@ -70,5 +70,8 @@ public static float ToMoisture(double RawReading)
             if (RawReading < 2.2) { return (float)((0.2632 * RawReading) - 0.0789); }
             return (float)((0.625 * RawReading) - 0.875);
         }
+
+        /// <summary> Does nothing. </summary>
+        public void EventTriggered(object Sender, EventArgs Event) { }
```
Comment (CaiB) [inline Components/Sensors/VH400.cs]: Why is this change in here? This reverts a wanted change now on `master`.

File: IO/BeagleBone/CANBusBBB.cs (hunk 3 above)
Comment (CaiB) [reply IO/BeagleBone/CANBusBBB.cs]: We could, but it's also not essential here. Up to you if you want to go through that effort.

File: IO/BeagleBone/CANBusBBB.cs [hunk 5]
```diff
@@ -135,14 +180,53 @@ internal CANBusBBB(string CanName)
...
+				unsafe
+				{
+					for (int i = 0; i < Frame.DataLength; i++) { Payload[i] = Frame.Data[i]; }
+				}
+				return new Tuple<uint, byte[]>(Frame.CANID, Payload);
 			}
-			return new Tuple<uint, byte[]>(Frame.CANID, Payload);
+		}
+
+		private byte DLCToLen(byte CanDLC)
```
Comment (CaiB) [inline IO/BeagleBone/CANBusBBB.cs]: Naming of this method. Is this a length? Then call it `Length`, not `Len`. Also comment as to what this does.

File: IO/BeagleBone/CANBusBBB.cs [hunk 6]
```diff
@@ -135,14 +180,53 @@ internal CANBusBBB(string CanName)
...
+		{
+			byte[] Dlc2Len = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 12, 16, 20, 24, 32, 48, 64 };
+			return Dlc2Len[CanDLC & 0x0F];
+		}
+
+		private byte LenToDLC(byte CanLen)
```
Comment (CaiB) [inline IO/BeagleBone/CANBusBBB.cs]: Same here. Assuming this is a length, name it as such. Also provide a comment as to what this is doing. What is `DLC`?

File: IO/BeagleBone/CANBusBBB.cs [hunk 7]
```diff
@@ -135,14 +180,53 @@ internal CANBusBBB(string CanName)
...
+					for (int i = 0; i < Frame.DataLength; i++) { Payload[i] = Frame.Data[i]; }
+				}
+				return new Tuple<uint, byte[]>(Frame.CANID, Payload);
 			}
-			return new Tuple<uint, byte[]>(Frame.CANID, Payload);
+		}
+
+		private byte DLCToLen(byte CanDLC)
+		{
+			byte[] Dlc2Len = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 12, 16, 20, 24, 32, 48, 64 };
```
Comment (CaiB) [inline IO/BeagleBone/CANBusBBB.cs]: Capitalization. Also this can probably be a `const`?

File: IO/BeagleBone/CANBusBBB.cs [hunk 8]
```diff
@@ -135,14 +180,53 @@ internal CANBusBBB(string CanName)
...
+			return Dlc2Len[CanDLC & 0x0F];
+		}
+
+		private byte LenToDLC(byte CanLen)
+		{
+			byte[] Len2DLC =
```
Comment (CaiB) [inline IO/BeagleBone/CANBusBBB.cs]: This could be a `const`.

File: IO/BeagleBone/CANBusBBB.cs [hunk 9]
```diff
@@ -152,15 +236,29 @@ internal CANBusBBB(string CanName)
...
-				CANFrame Frame = new CANFrame();
-				Frame.CANID = ID;
-				Frame.DataLength = (byte)Data.Length;
-				for (int i = 0; i < Data.Length; i++) { Frame.Data[i] = Data[i]; }
-				write(Socket, ref Frame, Marshal.SizeOf(Frame));
+				if (Extended)
+				{
+					ExtendedCANFrame Frame = new ExtendedCANFrame();
+					Frame.CANID = ID | CAN_EFF_FLAG;
+					Frame.DataLength = DLCToLen(LenToDLC((byte)Data.Length));
```
Comment (CaiB) [inline IO/BeagleBone/CANBusBBB.cs]: Those methods are really not well named, here it looks like you are doing something then undoing it.

File: IO/BeagleBone/CANBusBBB.cs (hunk 8 above)
Comment (save-buffer) [reply IO/BeagleBone/CANBusBBB.cs]: sounds good, doesn't work

File: IO/BeagleBone/CANBusBBB.cs (hunk 7 above)
Comment (save-buffer) [reply IO/BeagleBone/CANBusBBB.cs]: sounds good, doesn't work

**Pull request closed without merging**